### PR TITLE
Allow leading context parameters in extension methods

### LINKED ISSUE #9530: Support contextual dependent extensions
## Minimized example

We should support 
```Scala
trait Scope:
  type Expr

extension (using s: Scope)(expr: s.Expr)
  def show = "expr"

def f(using s: Scope)(s.Expr) = s.show
```

or the less performant

```scala
trait Scope:
  type Expr

given (using s: Scope) as AnyRef:
  extension (expr: s.Expr)
    def show = "expr"

def f(using s: Scope)(expr: s.Expr) = expr.show
```

## Expectation

One of those should be supported to be able to encode 
```scala
implicit class Ops(using s: Scope)(e: s.Expr):
  def show = "expr"
```
which does work and has several use cases.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #9530

### CODE REVIEW

**Comment by odersky** [comment]
@nicolasstucki If you have more examples where this would be useful, it would be good to try them out, and either use them in the library or add them as test cases.

**Comment by prolativ** [comment]
It would be good to also add some documentation for this feature.

I played a bit with this branch and it looks that when having code like this
```scala
object Foo
trait Bar
given bar: Bar with {}
extension (using Bar)(foo: Foo.type) def xxxx = 1
```
you cannot provide the using section explicitly when using extension syntax, e.g.
```scala
Foo.xxxx(using bar) // won't compile
```
Instead you need to fallback to a dotless method call
```scala
xxxx(using bar)(Foo)
```
Is this intentional? To me this seems somehow inconsistent as in case of both type parameters and postfix using clauses the parameters of an extension can be passed explicitly in all notations (even though type parameters in the definition of an extension are placed before the extension receiver parameter)
```scala
extension [A](foo: Foo.type)(using Bar) def yyyy = 1
val a = Foo.yyyy
val b = Foo.yyyy[Int](using bar)
val c = yyyy[Int](Foo)(using bar)
```

Changes introduced in this PR will require some adjustments to code completion as before we could assume that the first parameter of an extension method would always be the receiver, which will not necessarily be true anymore

**Comment by odersky** [comment]
> Is this intentional? To me this seems somehow inconsistent as in case of both type parameters and postfix using clauses the parameters of an extension can be passed explicitly in all notations (even though type parameters in the definition of an extension are placed before the extension receiver parameter)

Yes, that's intentional. Leading using clauses can be passed only if the method is called directly. The analogy with type parameters is unfortunate, but we plan to change the rules for type parameters so that type parameters can follow both `extension` and `def name`. In that case, the type parameters following `extension` also can be passed only in non-extension calls.
 

Who will take care of the changes in code completion?

**Comment by prolativ** [comment]
Ok, that makes sense. I can take care of code completion

File: docs/docs/reference/contextual/right-associative-extension-methods.md
Comment (prolativ) [inline docs/docs/reference/contextual/right-associative-extension-methods.md:26]: Shouldn't there be
```scala
val y = x; xs.+:(y)
```
instead of 
```scala
val y = x; xs +: y
```
?
And what if `x` is a by-name parameter?

File: docs/docs/reference/contextual/right-associative-extension-methods.md
Comment (prolativ) [inline docs/docs/reference/contextual/right-associative-extension-methods.md:26]: Sorry for commenting the same place twice but it was only now that I spotted it:
If `xs` is impure  but `x` is pure there should be no assignment, right? In general (im)purity of `xs` doesn't seem to play any role here. So it probably should be
```
if `x` is a pure expressions or a call-by-name parameter
```

Comment (odersky) [reply docs/docs/reference/contextual/right-associative-extension-methods.md:26]: Yes, we can make the change. I leave that to you since I am working on something else now.

Comment (prolativ) [reply docs/docs/reference/contextual/right-associative-extension-methods.md:26]: It looks like I can't push to this branch as I don't have permissions to dotty-staging. I didn't even notice that before as I created PRs from my own fork.

Comment (odersky) [reply docs/docs/reference/contextual/right-associative-extension-methods.md:26]: I see you were not added yet. I just invited you with push rights to dotty-staging.

Comment (prolativ) [reply docs/docs/reference/contextual/right-associative-extension-methods.md]: Thanks. Changes pushed

**Comment by prolativ** [comment]
@odersky I played a bit with this branch and it looks like leading using clauses make the search of extensions in companion objects and givens in scope not work.
According to the spec:
```
There are four possible ways for an extension method to be applicable:

1. The extension method is visible under a simple name, by being defined or inherited or imported in a scope enclosing the reference.
2. The extension method is a member of some given instance that is visible at the point of the reference.
3. The reference is of the form r.m and the extension method is defined in the implicit scope of the type of r.
4. The reference is of the form r.m and the extension method is defined in some given instance in the implicit scope of the type of r.
```
but if you look at this example
```scala
trait Food
case class Banana(color: String) extends Food

trait Diet[A <: Animal]:
  type F <: Food
  def food: Seq[F]

trait Animal
object Animal:
  extension [A <: Animal](using diet: Diet[A])(animal: A) def food1 = diet.food
  extension [A <: Animal](animal: A)(using diet: Diet[A]) def food2 = diet.food

extension [A <: Animal](using diet: Diet[A])(animal: A) def food3 = diet.food
extension [A <: Animal](animal: A)(using diet: Diet[A]) def food4 = diet.food

trait Monkey extends Animal

given Diet[Monkey] with
  type F = Banana
  def food: Seq[Banana] = Seq(new Banana("yellow"), Banana("green"))

trait FoodOps
given FoodOps with
  extension [A <: Animal](using diet: Diet[A])(animal: A) def food5 = diet.food
  extension [A <: Animal](animal: A)(using diet: Diet[A]) def food6 = diet.food


val monkey = new Monkey {}

val foods = Seq(
  monkey.food1, // doesn't compile
  monkey.food2, // compiles
  monkey.food3, // compiles
  monkey.food4, // compiles
  monkey.food5, // doesn't compile
  monkey.food6, // compiles
)
```
you'll see that for cases 2 and 3 from the spec this doesn't work as expected (as case 4 is kind of a composition of 2 and 3, I guess it doesn't work either). Passing the type parameter to `.foodN` doesn't help

**Comment by odersky** [comment]
@prolativ I see. This will be difficult to fix. I believe our best bet is to wait for #10950 to make it in. That will simplify handling to type arguments to extension methods and might already fix the problem. If not, we have a simpler base to work with.

**Comment by prolativ** [comment]
Allowing multiple type parameters sections doesn't seem to be going to change anything here. The problem here is that when an extension method is found in an companion object, it is discarded from the search because when the receiver type is compared to the first parameter of the extension, they don't match anymore because the first parameter can now be a using clause. But if the extension is available directly in the current scope a totally different mechanism is used to find it and this works. For extensions located inside implicits the search uses the same mechanism as for implicit conversions. I'm wondering whether there is any particular reason why we don't use the same mechanism to find extensions in all cases

**Comment by nicolasstucki** [comment]
> @nicolasstucki If you have more examples where this would be useful, it would be good to try them out, and either use them in the library or add them as test cases.

I added some tests and adapted some code in dokka to use extension methods on reflect symbols directly (this is only a first step). From the reflection use cases perspective it looks good.

**Comment by nicolasstucki** [comment]
@odersky #10950 will not help with this case as we cannot move the type parameter to the `def`. I would have expected the expected `food1` and `food2` to behave the same. Is there any reason why they should not?
```scala
trait Animal
object Animal:
  extension [A <: Animal](using diet: Diet[A])(animal: A) def food1 = diet.food
  ...
extension [A <: Animal](using diet: Diet[A])(animal: A) def food3 = diet.food

...
monkey.food1, // doesn't compile
...
monkey.food3, // compiles
...
```

File: community-build/src/scala/dotty/communitybuild/FieldsImpl.scala
Comment (nicolasstucki) [inline community-build/src/scala/dotty/communitybuild/FieldsImpl.scala:16]:
```suggestion
    val selects = symbols.map(Select(projectsTree, _).asExprOf[T])
```

File: compiler/src-non-bootstrapped/scala/quoted/runtime/impl/QuotesImpl.scala
Comment (nicolasstucki) [inline compiler/src-non-bootstrapped/scala/quoted/runtime/impl/QuotesImpl.scala:65]:
```suggestion
      if self.isExprOf[X] then
```

File: tests/pos/i9530.scala
Comment (prolativ) [inline tests/pos/i9530.scala:31]: Now everything should compile ;)
We should probably also add cases for extensions hidden in implicits which are themselves defined in a companion object

**Comment by nicolasstucki** [comment]
Rebased and resolved confilcts

File: library/src/scala/quoted/ExprMap.scala
Comment (nicolasstucki) [inline library/src/scala/quoted/ExprMap.scala:150]:
```suggestion
      .asExprOf[T]
```

File: tasty/src/dotty/tools/tasty/TastyFormat.scala
Comment (nicolasstucki) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:77]:
```suggestion

```

Comment (nicolasstucki) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:77]: Maybe move it to the description of `SPLITCLAUSE`

File: tasty/src/dotty/tools/tasty/TastyFormat.scala
Comment (nicolasstucki) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:74]: Here `same kind` is not really descriptive. One could think that this can be used to separate `(a: A)(b: B)` and `[A][B]` but not `(a: A)[B](c: C)` which appears internally in extensions.

Comment (odersky) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:74]: But that's exactly what it does. It separates `(a: A)(b: B)` and `[A][B]` but not `(a: A)[B](c: C)`.

Comment (nicolasstucki) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:74]: Ok. Then the awkward part would be that to encode `[A][B]` we would need a `TermParam`, though this would always be indirectly through a `Param`. For consistency, I would have expected `[A][B]` to be representable as a `TypeParam*`.

File: tasty/src/dotty/tools/tasty/TastyFormat.scala
Comment (odersky) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:77]: In fact, we should just remove this line.

File: tasty/src/dotty/tools/tasty/TastyFormat.scala
Comment (odersky) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:74]: No, TypeParam* really means a single list of type parameters. Constructors and classes will keep that restriction; it's not planned that they will allow curried type parameter lists for now.

**Comment by odersky** [comment]
@nicolasstucki Can you rebase and apply your suggestions? I tried to rebase bot got too much breakage relative to my branch to be comfortable with going ahead.

**Comment by nicolasstucki** [comment]
Yes, I'll rebase

File: tasty/src/dotty/tools/tasty/TastyFormat.scala
Comment (nicolasstucki) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:77]: Should we future proof this definition by allowing `class Foo[T](a: A)[U](a: B)` internally in TASTY?
```
  Template      = TEMPLATE       Length Param* parent_Term* Self?
```

Comment (odersky) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:77]: I think that's not on the agenda right now. I see all sorts of complications arising from this. 

So I think all of this can wait until we go to the next major redesign for Tasty in 3.T

**Review by nicolasstucki — APPROVED** [review]

**Comment by nicolasstucki** [comment]
@odersky can you double-check the commits I added and then merge?

**Comment by odersky** [comment]
@nicolasstucki Looks all good now. Thanks for your help!

**Pull request merged**
